discovery_engine_core/core/src/engine.rs

```diff
@@ -337,11 +338,15 @@ where
     pub async fn set_markets(
         &mut self,
         history: &[HistoricDocument],
-        markets: Vec<Market>,
+        new_markets: Vec<Market>,
     ) -> Result<(), Error> {
-        *self.config.markets.write().await = markets;
-
         {
+            let markets = &mut *self.config.markets.write().await;
+            let mut old_markets = replace(markets, new_markets);
+            let new_markets_set = markets.iter().collect::<HashSet<_>>();
+            old_markets.retain(|market| !new_markets_set.contains(market));
```

Comment:
❓ we could use the vector directly instead of collecting into a set before. this would mean linear instead of constant lookup time, but the number of markets is very limited and we save the set allocation.
```suggestion
            old_markets.retain(|market| !markets.contains(market));
```